### ISSUE TITLE
handling of #IND00 and #INF00

### ISSUE THREAD

**Issue by gaiaclary** [KhronosGroup/OpenCOLLADA#591]
Hi;
I got a report about a broken collada import. I found the problem is related to a series of lines containing:

    -1.#IND00` -1.#IND00 `-1.#IND00

OpenCollada actually sees the initial "-1." as a value of -1 but then reports the rest of the line as Parse error:

    Error: ERROR_TEXTDATA_PARSING_FAILED
    Element: float_array,
    Line: 178665, Column: 14, Additional: #IND00 -1.#IND00 -1.

I have seen the very same Collada file can be imported to other tools and game engines, although it is not clear to me how to process the lines properly. When we treat the parse errors just as warnings, then we get corrupted data in Blender, probably because the numbers(data counters) aren't right in that case.

I wonder if opencollada should(could) treat #IND00 and #INF00 as std::numeric_limits<T>::infinity() instead of creating a parsing error here. Any opinions ?

**Comment by lasalvavida** [comment]
Related: https://github.com/KhronosGroup/COLLADA2GLTF/issues/194

I don't think it's unreasonable to try to do something sane here. I think if anything, we should recognize that there is a whitespace separated token that failed to be parsed as a number, and write some default value into the resulting array (either NaN or zero).

I'm still going to do bounds checking in COLLADA2GLTF, but it would go a long way for OpenCOLLADA to emit a data array of the proper size when these errors occur.

**Comment by emackey** [comment]
`NaN` is not valid glTF or JSON, please don't write any from COLLADA2GLTF.

Sometimes these kinds of values can come from "harmless" sources, such as normal vectors of degenerate triangles, where it doesn't matter what data is filled in there.  But `NaN` is strongly disallowed in glTF, and will break most parsers.  I haven't checked COLLADA spec but I would expect `-1.#IND00` to be disallowed there as well.

If the user really wants to try to salvage a broken file, perhaps there could be an option to convert these values to zeros.  Zero-length normal vectors aren't valid glTF either, but, some viewers will still render with them.  The best solution though is to tell the user to clean up the source data.

**Comment by lasalvavida** [comment]
> NaN is not valid glTF or JSON, please don't write any from COLLADA2GLTF.

That's not what I meant. I was speaking strictly about the data array produced by OpenCOLLADA from parsing. This has nothing to do with COLLADA2GLTF or even glTF for that matter.

The problem is that when we consume this data, OpenCOLLADA produces an array of indices that may go past the bounds of the corresponding data array because elements have been dropped during parsing.

**Comment by gaiaclary** [comment]
I agree the best solution is to keep the data sane. Also OpenCollada actually reports those errors to the Importer's error handler. So it does not drop data silently, but it rings the warning bells. :) 

But that does not help a lot, because currently the error handler can only decide to go ahead (and let Blender crash because data is dropped) or abort the import entirely which is not much better from the user's point of view.

I also tested what happens when i blindly replace all occurrences of "1.#IND00" by "0" in the Collada file. The result can be safely imported into Blender. And i not even see any visual errors in the imported objects.  So at least the user is happy again :) 

Would it be an option to improve the error handler such that it allows to "replace the broken data" by something that the error handler decides? then we could in principle replace the IND00 values by whatever we think is best suited for our tool (let it be "1", "0", "NaN," ...) and let the parser go ahead with the fixed data... Does that make sense ?
